zipline/data/hdf5_daily_bars.py

```diff
@@ -430,6 +435,10 @@ def load_raw_arrays(self,
             (minutes in range, sids) with a dtype of float64, containing the
             values for the respective field over start and end dt range.
         """
+        self._validate_assets(assets)
```

Comment:
for `load_raw_arrays`, I could imagine just returning `nan` for the assets that don't exist. For code that wants to use this safely, they need to filter down the the known assets, which is hard to do because it is not part of the interface of a session bar reader. An example of a place where I ran into this was computing dividend ratios. When doing that, I needed to manually exclude any assets that we don't have pricing for, but if we just return all nan columns I can continue to vectorize the calculation and then strip nans in the final result.

cc @ssanderson what do you think about this?

Reply:
Looking at what the other BarReaders do here:

- InMemoryDailyBarReader raises a NoDataForSid on unexpected sids.
- BcolzDailyBarReader doesn't handle this case at all (looks like it would raise a KeyError from initial inspection).
- BcolzMinuteBarReader raises NoDataForSid
- The other readers all delegate to a sub-reader.

Given the current behavior, it seems like raising an error is the behavior that's most consistent with the rest of the system right now. I buy that in some contexts the nan-returning behavior might be preferable, but I don't think we actually have any use-cases where that's **required** right now, so the current erroring behavior also feels a bit safer to me (returning nans is useful if the caller knowingly gave us unknown sids, but it could also mask bugs where the caller expected us to have the requested sids). I also think that we should probably make `sids` part of the public interface for at least some of our readers, since that's a requirement for many natural composition patterns. Having readers expose their sids might lessen the value of the nan-returning behavior.